File: stack-agent-manager/frontend/types/stack.ts
import { UserInfo } from "./user";
import { AgentResponse } from "./agent";

export interface Stack {
  id: string;
  name: string;
  namespace: string | null;
  status: string;
  description: string | null;
  created_at: string;
  created_by: UserInfo;
  updated_at: string;
  updated_by: UserInfo;
}

export interface StackListResponse extends Stack {
  agent_count: number;
}

export interface StackDetailResponse extends Stack {
  agents: AgentResponse[];
}

export interface StackCreate {
  name: string;
  description?: string;
}

export interface StackUpdate {
  name?: string;
  description?: string;
}

export interface StackListPaginated {
  items: StackListResponse[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

File: stack-agent-manager/backend/app/models/stack.py
from datetime import datetime
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid

from app.database import Base


class Stack(Base):
    __tablename__ = "stacks"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False, index=True)
    namespace = Column(String(255), nullable=True, unique=True, index=True)
    description = Column(Text, nullable=True)
    status = Column(String(50), nullable=False, default="creating", index=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="RESTRICT"), nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_by = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="RESTRICT"), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    creator = relationship("User", foreign_keys=[created_by], back_populates="stacks_created")
    updater = relationship("User", foreign_keys=[updated_by], back_populates="stacks_updated")
    agents = relationship("Agent", back_populates="stack", cascade="all, delete-orphan")

